### PR TITLE
Minor: simplify DataSource statistics code

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

This is part of https://github.com/apache/arrow-datafusion/issues/8078

## Rationale for this change

I am trying to improve the handling of statistics by changing the internal representation, so I would like less to change. 

My longer term goal is to keep all the calculation and comparison code for `Statistics` within the `Statistics` class itself, which will make testing and changing implementation details

## What changes are included in this PR?

1. remove some direct manipulation of `Precision` and use pre-existing methods `min`, `max` and `add`

## Are these changes tested?
Covered by existing tests

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/core/src/datasource/statistics.rs [hunk 1]
```diff
@@ -211,49 +199,3 @@ pub(crate) fn get_col_stats(
...
-        }
-        (Precision::Exact(_), Precision::Absent) => max_values.to_inexact(),
-        (Precision::Absent, Precision::Exact(_)) => max_nominee.to_inexact(),
-        (Precision::Absent, Precision::Inexact(_)) => max_nominee,
-        (Precision::Absent, Precision::Absent) => Precision::Absent,
-        _ => max_values,
-    }
-}
-
-/// If the given value is numerically lesser than the original minimum value,
```
Comment (alamb) [inline datafusion/core/src/datasource/statistics.rs:238]: This is the same as https://github.com/apache/arrow-datafusion/blob/eef654c3b0c22b1f845b1441320b8bb718ddd605/datafusion/common/src/stats.rs#L92-L103

File: datafusion/core/src/datasource/statistics.rs [hunk 2]
```diff
@@ -211,49 +199,3 @@ pub(crate) fn get_col_stats(
         })
         .collect()
 }
-
-/// If the given value is numerically greater than the original maximum value,
-/// return the new maximum value with appropriate exactness information.
-fn set_max_if_greater(
```
Comment (alamb) [inline datafusion/core/src/datasource/statistics.rs:217]: This is the same as https://github.com/apache/arrow-datafusion/blob/eef654c3b0c22b1f845b1441320b8bb718ddd605/datafusion/common/src/stats.rs#L75-L87

File: datafusion/core/src/datasource/statistics.rs [hunk 3]
```diff
@@ -160,17 +159,6 @@ pub(crate) fn create_max_min_accs(
     (max_values, min_values)
 }
 
-fn add_row_stats(
```
Comment (alamb) [inline datafusion/core/src/datasource/statistics.rs:163]: This is the same as https://github.com/apache/arrow-datafusion/blob/eef654c3b0c22b1f845b1441320b8bb718ddd605/datafusion/common/src/stats.rs#L119-L126

**Comment by alamb** [comment]
cc @berkaysynnada  and @ozankabak

**Review by ozankabak — APPROVED** [review]
LGTM, thank you

**Review by viirya — APPROVED** [review]

File: datafusion/core/src/datasource/statistics.rs (hunk 2 above)
Comment (berkaysynnada) [reply datafusion/core/src/datasource/statistics.rs:217]: There is a similar difference here as well. Max values are conserved by relaxing the exactness when an absent statistic is read from the file.

File: datafusion/core/src/datasource/statistics.rs (hunk 1 above)
Comment (berkaysynnada) [reply datafusion/core/src/datasource/statistics.rs:238]: Same here

**Review by berkaysynnada — COMMENTED** [review]
The functions such as `add()`, `min()`, and `max()`, which I initially introduced in the precision API, were designed to operate in the safest way. However, I observe that the common usage of these functions in the code is trying to preserve statistical values across parameter updates by converting them to inexact values. What are your thoughts on modifying the functionalities of the precision methods instead? I mean these precision methods' logic could be replaced by the functions you removed.

Thanks for the simplifications

**Comment by ozankabak** [comment]
@berkaysynnada, good catch! Let's consider changing the `add`/`max`/`min` methods to create `Inexact` types when they encounter an `Absent`. @alamb, can you think of a case where we would need to propagate `Absent` if any operand is `Absent`?

**Comment by alamb** [comment]
> What are your thoughts on modifying the functionalities of the precision methods instead? I mean these precision methods' logic could be replaced by the functions you removed.

This is an excellent catch @berkaysynnada  -- thank you. My thoughts are that I would like to initially not change any behavior unless it is clearly a bug. So in this case, I will:

1.  add tests to the datasource statistics calculations so there is coverage for this behavior
2. move this code into `Precision` (as a method like `estimated_add` or something) 

The reason to move the code into `Precision` is to consolidate how this structure is being used (otherwise we can end up with surprises like this)

> @berkaysynnada, good catch! Let's consider changing the add/max/min methods to create Inexact types when they encounter an Absent. @alamb, can you think of a case where we would need to propagate Absent if any operand is Absent?

I can not. One test would be to try to change `Precision::add` and see if any tests break, though given our lack of coverage maybe this doesn't tell us as much as we would like.

File: datafusion/core/src/datasource/statistics.rs (hunk 3 above)
Comment (alamb) [inline datafusion/core/src/datasource/statistics.rs:163]: It is interesting that there is very similar code in https://github.com/apache/arrow-datafusion/blob/a892300a5a56c97b5b4ddc9aa4a421aaf412d0fe/datafusion/core/src/datasource/file_format/parquet.rs#L503-L525 that uses `add` 🤔

**Review by alamb — COMMENTED** [review]
How embarassing -- I even noted the lack of unit tests in this file on https://github.com/apache/arrow-datafusion/pull/7793/files#r1358805175 🤦

File: datafusion/core/src/datasource/statistics.rs (hunk 3 above)
Comment (berkaysynnada) [reply datafusion/core/src/datasource/statistics.rs:163]: It seems like it is always adding exact values, so there should be no difference on the results whether it is used `add()` or `estimated_add()`

**Comment by berkaysynnada** [comment]
> How embarassing -- I even noted the lack of unit tests in this file on https://github.com/apache/arrow-datafusion/pull/7793/files#r1358805175 🤦

Sadly, while I was working on the changes, I noticed that in many parts where we use statistics, we don't test them enough. You have done really well starting the Epic. It definitely requires quite a bit of work.

**Comment by alamb** [comment]
> Improve statistics test coverage #8228

Yes, thank you -- I plan to improve the situation via https://github.com/apache/arrow-datafusion/issues/8228

**Comment by alamb** [comment]
I'll keep working on statistics under the aegis of a different issue

**Pull request closed without merging**